// @openzeppelin/defender-sdk-account-client@2.7.0/lib/api/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountClient = void 0;
const defender_sdk_base_client_1 = require("@openzeppelin/defender-sdk-base-client");
const account_1 = require("../models/account");
const PATH = '/account';
const API_KEY_PATH = '/api-keys';
class AccountClient extends defender_sdk_base_client_1.BaseApiClient {
    getPoolId() {
        return process.env.DEFENDER_POOL_ID ?? 'us-west-2_94f3puJWv';
    }
    getPoolClientId() {
        return process.env.DEFENDER_POOL_CLIENT_ID ?? '40e58hbc7pktmnp9i26hh5nsav';
    }
    getApiUrl() {
        return process.env.DEFENDER_API_URL ?? 'https://defender-api.openzeppelin.com/';
    }
    async getUsage(params) {
        const searchParams = new URLSearchParams({
            ...(params?.quotas && { quotas: params.quotas.join(',') }),
            ...(params?.date && { date: new Date(params.date).toISOString() }),
        });
        return this.apiCall(async (api) => api.get(`${PATH}/usage?${searchParams.toString()}`));
    }
    async listApiKeyCapabilities() {
        const res = await this.apiCall(async (api) => api.get(`${API_KEY_PATH}/${this.apiKey}/capabilities`));
        return res.map(account_1.toApiKeysCapabilityV2);
    }
}
exports.AccountClient = AccountClient;
