// @openzeppelin/defender-sdk-relay-signer-client@2.7.0/lib/ethers/provider-v5.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefenderRelayProviderV5 = void 0;
const providers_1 = require("@ethersproject/providers");
const signer_v5_1 = require("./signer-v5");
const properties_1 = require("@ethersproject/properties");
const bignumber_1 = require("@ethersproject/bignumber");
const relayer_1 = require("../relayer");
const api_1 = require("../api");
class DefenderRelayProviderV5 extends providers_1.StaticJsonRpcProvider {
    constructor(credentials) {
        super((0, api_1.getApiUrl)());
        this.credentials = credentials;
        this.relayer = new relayer_1.Relayer(credentials);
    }
    async detectNetwork() {
        if (this.network != null) {
            return this.network;
        }
        // Logic from JsonRpcProvider.detectNetwork
        let chainId = null;
        try {
            chainId = await this.send('eth_chainId', []);
        }
        catch (error) {
            try {
                chainId = await this.send('net_version', []);
            }
            catch (error) {
                // Key difference from JsonRpcProvider.detectNetwork logic
                // This surfaces error to caller (like QuotaExceeded) instead of squashing it
                throw error;
            }
        }
        if (chainId === null) {
            throw new Error('could not detect chainId');
        }
        // Logic from JsonRpcProvider.detectNetwork
        const getNetwork = (0, properties_1.getStatic)(this.constructor, 'getNetwork');
        const network = getNetwork(bignumber_1.BigNumber.from(chainId).toNumber());
        if (!network) {
            throw new Error('could not detect network');
        }
        // Logic from StaticJsonRpcProvider.detectNetwork
        if (this._network == null) {
            (0, properties_1.defineReadOnly)(this, '_network', network);
            this.emit('network', network, null);
        }
        return network;
    }
    async send(method, params) {
        const request = { method, params };
        this.emit('debug', { action: 'request', request, provider: this });
        try {
            const result = await this.relayer.call(request);
            this.emit('debug', { action: 'response', request, response: result, provider: this });
            if (result.error) {
                const error = new Error(result.error.message);
                error.code = result.error.code;
                error.data = result.error.data;
                throw error;
            }
            return result.result;
        }
        catch (error) {
            this.emit('debug', { action: 'response', error, request: request, provider: this });
            throw error;
        }
    }
    getSigner() {
        return new signer_v5_1.DefenderRelaySignerV5(this.relayer, this, {});
    }
}
exports.DefenderRelayProviderV5 = DefenderRelayProviderV5;
