// @openzeppelin/defender-sdk-address-book-client@2.7.0/lib/api/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddressBookClient = void 0;
const defender_sdk_base_client_1 = require("@openzeppelin/defender-sdk-base-client");
const PATH = '/address-book';
class AddressBookClient extends defender_sdk_base_client_1.BaseApiClient {
    getPoolId() {
        return process.env.DEFENDER_POOL_ID ?? 'us-west-2_94f3puJWv';
    }
    getPoolClientId() {
        return process.env.DEFENDER_POOL_CLIENT_ID ?? '40e58hbc7pktmnp9i26hh5nsav';
    }
    getApiUrl() {
        return process.env.DEFENDER_API_URL ?? 'https://defender-api.openzeppelin.com/';
    }
    async list() {
        return this.apiCall(async (api) => api.get(`${PATH}`));
    }
    async create(address) {
        return this.apiCall(async (api) => api.post(`${PATH}`, address));
    }
    async delete(addressBookId) {
        const encodedId = encodeURIComponent(addressBookId);
        return this.apiCall(async (api) => api.delete(`${PATH}/${encodedId}`));
    }
}
exports.AddressBookClient = AddressBookClient;
