// @openzeppelin/defender-sdk-key-value-store-client@2.7.0/lib/local.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyValueStoreLocalClient = void 0;
const fs_extra_1 = require("fs-extra");
/**
 * Uses a local file for representing the store as a JSON. All fs operations
 * are synchronous to prevent race conditions on put/del operations.
 */
class KeyValueStoreLocalClient {
    constructor(params) {
        this.path = params.path;
    }
    async get(key) {
        return this.getStore()[key];
    }
    async put(key, value) {
        this.updateStore((store) => {
            store[key] = value;
        });
    }
    async del(key) {
        this.updateStore((store) => {
            delete store[key];
        });
    }
    getStore() {
        return (0, fs_extra_1.readJsonSync)(this.path, { throws: false }) ?? {};
    }
    updateStore(updater) {
        const store = this.getStore();
        updater(store);
        (0, fs_extra_1.ensureFileSync)(this.path);
        (0, fs_extra_1.writeJsonSync)(this.path, store, { spaces: 2 });
    }
}
exports.KeyValueStoreLocalClient = KeyValueStoreLocalClient;
